linkerd/proxy/http/src/server.rs

```diff
@@ -60,137 +55,113 @@ impl<F, H> NewServeHttp<F, H> {
+    fn poll_ready(&mut self, _: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
         Poll::Ready(Ok(()))
     }
 
     fn call(&mut self, io: PrefixedIo<I>) -> Self::Future {
-        match self.clone() {
-            Self::Http {
-                version,
-                service,
-                drain,
-                mut server,
-            } => Box::pin(async move {
-                debug!(?version, "Handling as HTTP");
-                let (svc, closed) = SetClientHandle::new(io.peer_addr()?, service);
-                match version {
-                    Version::Http1 => {
-                        // Enable support for HTTP upgrades (CONNECT and websockets).
-                        let mut conn = server
-                            .http1_only(true)
-                            .serve_connection(io, upgrade::Service::new(svc, drain.clone()))
-                            .with_upgrades();
-
-                        tokio::select! {
-                            res = &mut conn => {
-                                debug!(?res, "The client is shutting down the connection");
-                                res?
-                            }
-                            shutdown = drain.signal() => {
-                                debug!("The process is shutting down the connection");
-                                Pin::new(&mut conn).graceful_shutdown();
-                                shutdown.release_after(conn).await?;
-                            }
-                            () = closed => {
-                                debug!("The stack is tearing down the connection");
-                                Pin::new(&mut conn).graceful_shutdown();
-                                conn.await?;
-                            }
+        let Self {
+            version,
+            inner,
+            drain,
+            mut server,
+        } = self.clone();
+        debug!(?version, "Handling as HTTP");
+
+        Box::pin(async move {
+            let (svc, closed) = SetClientHandle::new(io.peer_addr()?, inner.clone());
+
+            match version {
+                Version::Http1 => {
+                    // Enable support for HTTP upgrades (CONNECT and websockets).
+                    let mut conn = server
+                        .http1_only(true)
+                        .serve_connection(io, upgrade::Service::new(svc, drain.clone()))
+                        .with_upgrades();
+                    tokio::select! {
+                        res = &mut conn => {
+                            debug!(?res, "The client is shutting down the connection");
+                            res?
+                        }
+                        shutdown = drain.signal() => {
+                            debug!("The process is shutting down the connection");
+                            Pin::new(&mut conn).graceful_shutdown();
+                            shutdown.release_after(conn).await?;
+                        }
```

Comment:
it seems like it *could* be possible to factor this out into a drain layer that's a superset of this logic (actually do a graceful shutdown) and `NewRetain` (just hold the shutdown alive). but, there would basically just be two of these (one for `ServeHttp` and the other for TCP), so it's probably not worth it. it just seems weirdly asymmetrical for the TCP stack to use a separate layer for drains and the HTTP stack to have it tightly coupled.

please feel free to ignore this one, it's not important! :)